### PR TITLE
Type simplification: get rid of TypeDefiner, Create(Internal)Definitions 

### PR DESCRIPTION
**Pull request by Porges**
1. Replace all uses of `TypeDefiner` interface with `TypeDefinition`. This removes:
 * `EnumDefinition`
 * `ResourceDefinition`
 * `StructDefinition`

2. Add `AsDeclarations` to `Type` interface. There is a helper `SimpleAsDeclarations` for straightforward type definitions. For enums/structs/resources this is the code that used to live in the “Definition” types.

3. Remove `CreateDefinitions` (replaced by `MakeTypeDefinition`) and `CreateInternalDefinitions` from `Type` interface. The work that `CreateInternalDefinitions` used to do is now run as a completely separate pipeline stage `"name inner types for CRD"` so it is clearer what it is doing (we can even turn it off, but `controller-gen` doesn’t like it!)

### CODE REVIEW

File: hack/generator/pkg/codegen/code_generator.go [hunk 1]
```diff
@@ -150,17 +145,35 @@ func (loader *cancellableJSONLoader) LoaderFactory() gojsonschema.JSONLoaderFact
...
-	}
+func loadSchema(ctx context.Context, idFactory astmodel.IdentifierFactory, configuration *config.Configuration) PipelineStage {
+	source := configuration.SchemaURL
 
-	schema, err := sl.Compile(loader)
-	if err != nil {
-		return nil, errors.Wrapf(err, "error loading schema from %q", source)
-	}
+	return PipelineStage{
+		Name: fmt.Sprintf("load and walk schema from %q", source),
```
Comment (Porges) [inline hack/generator/pkg/codegen/code_generator.go:152]:
```suggestion
		Name: fmt.Sprintf("Load and walk schema from %q", source),
```

File: hack/generator/pkg/astmodel/enum_type.go [hunk 2]
```diff
@@ -29,25 +30,98 @@ func NewEnumType(baseType *PrimitiveType, options []EnumValue) *EnumType {
...
+
+	validationComment := GenerateKubebuilderComment(enum.CreateValidation())
+	declaration.Doc.List = append(
+		declaration.Doc.List,
+		&ast.Comment{Text: "\n" + validationComment})
+
+	return declaration
+}
+
+func (enum *EnumType) createValueDeclaration(value EnumValue, name *TypeName) ast.Spec {
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/astmodel/enum_type.go:86]: I think these parameters should be the other way around ... maybe.

File: hack/generator/pkg/codegen/pipeline_name_types_for_crd.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+// nameTypesForCRD - for CRDs all inner enums and objects must be named, so we do it here
+func nameTypesForCRD(idFactory astmodel.IdentifierFactory) PipelineStage {
+
+	return PipelineStage{
+		Name: "Name inner types for CRD",
+		Action: func(ctx context.Context, types Types) (Types, error) {
+
+			result := make(Types)
+
+			// this is a little bit of a hack, better way to do it?
```
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/codegen/pipeline_name_types_for_crd.go:23]: it works fine so it doesn’t really _need_ improvement. just if someone thinks of a better way :)

File: hack/generator/pkg/astmodel/property_definition.go [hunk 4]
```diff
@@ -35,13 +35,13 @@ func NewPropertyDefinition(propertyName PropertyName, jsonName string, propertyT
 	}
 }
 
-// NewEmbeddedStructDefinition is a factory method for defining an embedding
-// of another struct type.
-func NewEmbeddedStructDefinition(structType Type) *PropertyDefinition {
+// NewEmbeddedDefinition is a factory method for defining an embedding
+// of another type.
+func NewEmbeddedDefinition(other Type) *PropertyDefinition {
```
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/astmodel/property_definition.go:40]: did we drop this from allOf?

File: hack/generator/pkg/codegen/pipeline_name_types_for_crd.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+					result[typeName] = typeDef
+				}
+			}
+
+			return result, nil
+		},
+	}
+}
+
+func handleType(def astmodel.TypeDefinition, idFactory astmodel.IdentifierFactory, getDescription func(*astmodel.TypeName) *string) []astmodel.TypeDefinition {
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline_name_types_for_crd.go]: Maybe a clearer name would be `nameInnerTypes`?

File: hack/generator/pkg/codegen/pipeline_stage.go [hunk 6]
```diff
@@ -7,12 +7,16 @@ package codegen
 
 import (
 	"context"
+
 	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
 )
 
+// Types is the set of all types being generated
+type Types = map[astmodel.TypeName]astmodel.TypeDefinition
```
Comment (babbageclunk) [inline hack/generator/pkg/codegen/pipeline_stage.go:15]: I think this should be just `type Types map[astmodel.TypeName]astmodel.TypeDefinition` - the `=` form is more meant for aliasing types from other packages to support gradual refactoring in a large codebase.

File: hack/generator/pkg/astmodel/enum_type.go (hunk 2 above)
Comment (Porges) [reply hack/generator/pkg/astmodel/enum_type.go:86]: make s sense

File: hack/generator/pkg/codegen/pipeline_stage.go (hunk 6 above)
Comment (Porges) [reply hack/generator/pkg/codegen/pipeline_stage.go:15]: I thought that that stopped you being able to use `range` iteration but it looks like it doesn’t. Will change 👍

File: hack/generator/go.sum [hunk 7]
```diff
@@ -13,7 +13,9 @@ github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/NYTimes/gziphandler v0.0.0-20170623195520-56545f4a5d46/go.mod h1:3wb06e3pkSAbeQ52E9H9iFoQsEEwGN64994WTCIhntQ=
 github.com/OneOfOne/xxhash v1.2.2/go.mod h1:HSdplMjZKSmBqAxg5vPj2TmRDmfkzw+cTzAElWljhcU=
+github.com/PuerkitoBio/purell v1.0.0 h1:0GoNN3taZV6QI81IXgCbxMyEaJDXMSIjArYBCYzVVvs=
```
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/go.sum:16]: I think they must be a dependency of something; `go mod tidy` does nothing about it.

[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/go.sum:6]: Figured it out: it was the generated code in a subdirectory causing them to be imported!

File: hack/generator/pkg/astmodel/type.go [hunk 8]
```diff
@@ -23,14 +24,88 @@ type Type interface {
...
+		return tv.VisitResourceType(tv, it, ctx)
+	}
+
+	panic(fmt.Sprintf("unhandled type: (%T) %v", t, t))
+}
 
-	// CreateInternalDefinitions creates definitions for nested types where needed (e.g. nested anonymous enums, structs),
-	// and returns the new, updated type to use in this type’s place.
-	CreateInternalDefinitions(nameHint *TypeName, idFactory IdentifierFactory) (Type, []TypeDefiner)
+func MakeTypeVisitor() TypeVisitor {
```
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/astmodel/type.go:86]: I did think of this and did it initially but it was a bit messy. It still runs very quickly 🙂

File: hack/generator/pkg/astmodel/package_definition.go [hunk 9]
```diff
@@ -98,23 +94,20 @@ func anyReferences(defs []TypeDefiner, defName *TypeName) bool {
 	return false
 }
 
-func partitionDefinitions(definitions []TypeDefiner) (resourceStructs []*ResourceDefinition, otherDefinitions []TypeDefiner) {
-
-	var resources []*ResourceDefinition
-	var notResources []TypeDefiner
+func partitionDefinitions(definitions []TypeDefinition) (resources []TypeDefinition, otherDefinitions []TypeDefinition) {
```
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/astmodel/package_definition.go]: replaced with a struct

File: hack/generator/pkg/astmodel/property_definition.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/astmodel/property_definition.go:40]: done

File: hack/generator/pkg/astmodel/type.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/astmodel/type.go:86]: I'll put this in as a TODO if we need it in future to be faster

File: hack/generator/pkg/astmodel/type.go [hunk 10]
```diff
@@ -23,14 +24,88 @@ type Type interface {
...
-	// that also must be included in the output.
-	CreateDefinitions(name *TypeName, idFactory IdentifierFactory) (TypeDefiner, []TypeDefiner)
+func (tv *TypeVisitor) Visit(t Type, ctx interface{}) Type {
+	if t == nil {
+		return nil
+	}
+
+	switch it := t.(type) {
+	case *TypeName:
+		return tv.VisitTypeName(tv, it, ctx)
```
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/astmodel/type.go:64]: I'm going to leave this, I think the context (ha) makes this pretty clear. I added a note about the use of the argument.

File: hack/generator/pkg/astmodel/package_definition.go [hunk 11]
```diff
@@ -98,23 +94,20 @@ func anyReferences(defs []TypeDefiner, defName *TypeName) bool {
...
+		if _, ok := def.Type().(*ResourceType); ok {
+			resources = append(resources, def)
 		} else {
-			notResources = append(notResources, def)
+			otherDefinitions = append(otherDefinitions, def)
 		}
 	}
 
-	return resources, notResources
+	return
```
[the remark this answers is not in the document]
Comment (Porges) [reply hack/generator/pkg/astmodel/package_definition.go]: goneburgers

**Pull request merged**